### PR TITLE
I206 User / contact updates not shown in UI

### LINKED ISSUE #206: Updates to user / contacts are not seen in UI
###  Context

Whenever the user updates their information or changes a contact, the information is successfully updated within the database, but the UI does not update to show the new state. This is because the information is stored in an atom that is unaware of the client state invalidation and does not listen for the database refetch.

### Expected Behavior

Contact / user updates should be shown within the client UI, this will make UX much better.

### PR DESCRIPTION
**Pull request by scratchclaggy**
## Change Summary

This PR addresses #206 and brings a number of improvements. One of the reasons that the previous bug was present was due to the fact that the user was treated as a contact, and this required some workarounds. This PR refactors things so that the contacts page is split like so:

```bash
src/pages/contact
├── [id]
│   ├── edit.tsx
│   └── index.tsx
├── index.tsx
└── new.tsx

1 directory, 4 files

---

src/pages/profile
├── edit.tsx
└── index.tsx

0 directories, 2 files
```

Now, instead of using multiple pieces of client side state (via Jotai) we can navigate to the profile page for the user and to contacts with dynamic page routes. Additionally, this also gives the ability to show the most newly created contact when adding a new contact, instead of redirecting to the contact list like before.

### Change Form
- [X] The pull request title has an issue number
- [X] The change works by "Smoke testing" or quick testing
- [ ] The change has tests
- [ ] The change has documentation

# Related Issue

- Resolve #206

### CODE REVIEW

**Comment by scratchclaggy** [comment]
TODO:

- New contact needs a back button
- Use page navigation for edit vs detail view

**Comment by dct0** [comment]
> * New contact needs a back button

In #222 I have made the cancel button present for both editing and adding a new contact. Having the 'back' button at the bottom of the page and requiring scrolling is not good design. Couldn't find a nice way to have the button at the top so maybe this could be made into a separate issue?

**Comment by scratchclaggy** [comment]
Should be ready for review now

File: firebase/functions/package.json [hunk 1]
```diff
@@ -23,10 +23,12 @@
   },
   "main": "lib/index.js",
   "dependencies": {
+    "@firebase/app-types": "0.x",
     "@google-cloud/billing": "3.1.3",
     "@google-cloud/pubsub": "3.2.1",
     "firebase-admin": "11.3.0",
-    "firebase-functions": "4.2.0"
+    "firebase-functions": "4.2.0",
+    "jest": "28.0.0"
```
Comment (dct0) [inline firebase/functions/package.json:31]: Should we keep this?

File: src/components/Contact/contactform.tsx [hunk 2]
```diff
@@ -176,21 +160,13 @@ const ContactForm = () => {
...
-              onClick={() => {
-                if (setIsEditing !== undefined) setIsEditing(false)
-              }}
-            >
-              Cancel
-            </Button>
-          )}
+          <Button intent='secondary' fullwidth onClick={() => router.back()}>
+            Cancel
+          </Button>
```
Comment (dct0) [inline src/components/Contact/contactform.tsx:169]: The default type for buttons is 'submit' so hitting cancel submits the form.

**Review by dct0 — CHANGES_REQUESTED** [review]
Good stuff! Just some small changes to make.

File: src/pages/contact/[id]/edit.tsx [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+          size='medium'
+          onClick={() => {
+            router.back()
+          }}
+        >
+          Back
+        </Button>
+      </div>
+      <ContactForm contact={contact} mutate={mutateContacts} />
+    </div>
```
[the remark this answers is not in the document]
Comment (scratchclaggy) [reply src/pages/contact/[id]/edit.tsx]: Yeah good idea, I've added it now.

**Review by dct0 — APPROVED** [review]
All good

**Pull request merged**